File: 백준/Gold/20040. 사이클 게임/사이클 게임.cc
/*
사이클 판별 -> union find

연결하려는 두 점이 서로 같은 집합이라면 사이클을 생성한다.

*/
#include <iostream>
using namespace std;
int parent[500000];
int find_root(int x) {
	if (parent[x] == x) return x;
	return parent[x] = find_root(parent[x]);
}
bool is_same_set(int x, int y) {
	int rx = find_root(x);
	int ry = find_root(y);
	if (rx == ry) return true;
	parent[ry] = rx;
	return false;
}

int main() {
	int n, m;// n개의 점, m번째 차례
	cin >> n >> m;
	
	for (int i = 0; i < n; i++) parent[i] = i;
	
	for (int i = 0; i < m; i++) {
		int x, y;
		cin >> x >> y;
		if (is_same_set(x, y)) {
			cout << i + 1 << "\n";
			return 0;
		}
	}
	cout << 0 << "\n";
	return 0;
}

File: 백준/Gold/20040. 사이클 게임/사이클 게임.py
import sys
n, m = map(int, input().split())

parent = [i for i in range(n)]

def find_root(x):
    if parent[x] == x:
        return x
    
    parent[x] = find_root(parent[x])
    return parent[x]

def union_set(x, y):
    x = find_root(x)
    y = find_root(y)

    if (x > y):
        parent[x] = y
    else:
        parent[y] = x
    return

def is_same_set(x, y):
    x = find_root(x)
    y = find_root(y)
    return x == y

for i in range(1, m + 1):
    a,b = map(int, input().split())
    
    if (is_same_set(a,b)):
        print(i)
        sys.exit(0)

    union_set(a,b)

print(0)

